drivers/pwm/pwm_led_esp32.c

```diff
@@ -0,0 +1,517 @@
+/*
+ * Copyright (c) 2017 Vitor Massaru Iha <vitor@massaru.org>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/* Include esp-idf headers first to avoid redefining BIT() macro */
+#include <esp_intr_alloc.h>
+#include <soc/dport_reg.h>
+#include <rom/gpio.h>
+#include <soc/gpio_sig_map.h>
+#include <soc/ledc_reg.h>
+
+#include <soc.h>
+#include <errno.h>
+#include <board.h>
+#include <pwm.h>
+#include <kernel.h>
+#include <gpio.h>
+#include <string.h>
+
+#define PWM_ESP32_HSCH_HPOINT(i) (LEDC_HSCH0_HPOINT_REG + (0x14 * i))
+#define PWM_ESP32_HSCH_DUTY(i) (LEDC_HSCH0_DUTY_REG + (0x14 * i))
+#define PWM_ESP32_HSCH_CONF0(i) (LEDC_HSCH0_CONF0_REG + (0x14 * i))
+#define PWM_ESP32_HSCH_CONF1(i) (LEDC_HSCH0_CONF1_REG + (0x14 * i))
+#define PWM_ESP32_HSTIMER(i) (LEDC_HSTIMER0_CONF_REG + (0x8 * i))
+
+#define PWM_ESP32_LSCH_HPOINT(i) (LEDC_LSCH0_HPOINT_REG + (0x14 * i))
+#define PWM_ESP32_LSCH_DUTY(i) (LEDC_LSCH0_DUTY_REG + (0x14 * i))
+#define PWM_ESP32_LSCH_CONF0(i) (LEDC_LSCH0_CONF0_REG + (0x14 * i))
+#define PWM_ESP32_LSCH_CONF1(i) (LEDC_LSCH0_CONF1_REG + (0x14 * i))
+#define PWM_ESP32_LSTIMER(i) (LEDC_LSTIMER0_CONF_REG + (0x8 * i))
+
+enum {
+	PWM_LED_ESP32_REF_TICK_FREQ,
+	PWM_LED_ESP32_APB_CLK_FREQ,
+};
+
+enum {
+	PWM_LED_ESP32_HIGH_SPEED,
+	PWM_LED_ESP32_LOW_SPEED
+};
+
+struct pwm_led_esp32_timer {
+	u8_t bit_num;
+	int freq;
+};
```

Comment:
Nitpick: there's a 3-byte hole here due to alignment. Since this struct is not in a fast path, it might be a good idea to use a packed attribute to reduce its size (the timer_cfg matrix in pwm_led_esp32_config will shrink from 64 to 40 bytes).